### DISCUSSION TITLE
Interfaz del KnockKnock Server

### DISCUSSION THREAD

**Discussion by cgrodriguezr** [hackatomik/SISCOx256#13] [General | Lab]
La pregunta directa es ¿Existirá alguna forma de darle una interfaz grafica al programa del KnockKnock Server? ¿Cual recomendarían para proyectos similares en Java?.

Ejecuntándolo en consola me di cuenta de que se me hacía incómodo tener que abrir dos pestañas de la terminal e ir cambiando de una pestaña a otra. Por eso se me ocurrió buscar algún framework o librería que me permitiera darle un apartado visual, pero no estoy seguro de cuál utilizar o tan siquiera si alguno será compatible con el programa.

**Comment by tapilew** [comment]
¡Buena pregunta!

Existen muchas formas. Con tal que ejecutes código que se pueda comunicar con el protocolo (TCP Sockets) del servidor (así como el ejemplo de KnockKnockClient.java), con cualquier tipo de interfaz o lenguaje de programación debería ser posible.

Hay frameworks de Java que permiten crear interfaces gráficas en desktop, pero actualmente lo más recomendado son tecnologías web (HTML, CSS, JavaScript). Con JavaScript se puede programar tanto para el navegador, desktop, mobile e incluso servidores; es demasiado versátil. De hecho yo estoy desarrollando la interfaz web del cliente con React (una librería de JavaScript para interfaces).

Si también quieres usar Java para la GUI (interfaz de usuario gráfica) en apps desktop, existen opciones como JavaFX o Java Swing. Habitualmente en la UTP piden usar Java Swing cuando enseñan Java pero en realidad JavaFX es "más moderno". Y si quieres consumirlo mediante una app de Android Java también sirve para eso.

**Reply by JamSwiftDev** [reply]
justamente iba a decir lo mismo y recomendar lo mismo.

**Comment by quantumquirkxyz** [comment]
Sí, es posible darle una interfaz gráfica al programa del KnockKnock Server utilizando varias opciones disponibles en Java. Si buscas una solución moderna y robusta, JavaFX es una excelente opción. JavaFX ofrece una amplia gama de componentes gráficos y permite diseñar interfaces atractivas y avanzadas. Proporciona herramientas para la creación de interfaces con diseño profesional y la capacidad de manejar animaciones y efectos visuales. Además, permite usar archivos FXML para simplificar el diseño de la interfaz. Alternativamente, Swing es una opción más tradicional y parte del JDK estándar, adecuada para interfaces más básicas pero funcionales. SWT (Standard Widget Toolkit), proporcionado por Eclipse, utiliza componentes nativos del sistema operativo y puede ofrecer una experiencia más cercana a las aplicaciones nativas. JGoodies es otra opción que mejora Swing con componentes adicionales y mejores prácticas para el diseño. Para empezar con JavaFX, necesitarás instalar el SDK de JavaFX, configurar tu entorno de desarrollo, y diseñar la interfaz gráfica utilizando herramientas como Scene Builder o directamente en código JavaFX, adaptando el programa del KnockKnock Server para interactuar con la GUI en lugar de la consola. Esta transición a una interfaz gráfica hará que la gestión y visualización de las funcionalidades del servidor sea mucho más intuitiva y cómoda.

**Reply by tapilew** [reply]
shhhh, JS es más cool
